### ISSUE TITLE
Goto definition problem

### ISSUE THREAD

**Issue by bobbrow** [microsoft/vscode-cpptools#2281]
Type: LanguageService
Moved from #1163 
@Flamefire said:

I have the same (or very similar) problem (cpptools + CMake tools):
- Open a CMake project & Configure with CMake tools
- Set `compileCommands` to the `compile_commands.json`
- Add `...foo/include` to `browse.path` (Shows up as `Folder: ...foo/include/ will be indexed`)

Results:
- Navigate to include file included with `#include <foo.hpp>` -> Works (after `browse.path`)
- Hover over function from a class in `foo.cpp` -> Works (function declaration shown)
- Click `Go to declaration` -> Not working (`No definition found`)

Edit: Seems to not only apply to all classes. I have another class from an include file in the same directory and there it works. But for the other I can't even go to the class itself. Only difference I could find is the usage of a member function template in the not-found class.   
Removing that function does not change the missing definition of the class BUT I get an error squiggle on usage of that function

**Comment by bobbrow** [comment]
@bobbrow said:

@Flamefire, can you provide a small example where this is the case?  Note: that if you're trying to do goto def on a symbol scoped by a namespace alias (and maybe also a typedef), this doesn't work currently.  [0.17.7-insiders](https://github.com/Microsoft/vscode-cpptools/releases/tag/v0.17.7-insiders) has a fix for the namespace alias issue.

For example:
```cpp
namespace N {
    class Foo {};
}
namespace Alias = N;
Alias::Foo f;  // GotoDef/Decl on 'Foo' doesn't work in 0.17.6
```

Many other GotoDef/Decl cases should be fixed with #255 which is one of the next big items on our backlog.

**Comment by bobbrow** [comment]
@FlameFire said:

Not really, the code I'm using is not open source (yet) and I can't reduce this. I'll try to describe what I can:

Work dir: ~/git/project/subdir/
Build dir: ~/build/project
Compile command (from compile_commands.json): `"/usr/bin/g++-6  -DBOOST_VARIANT_DO_NOT_USE_VARIADIC_TEMPLATES -I~/git/project/subdir/component/.. -isystem ~/foo/include   -Wall -Wextra -Werror -O3 -DNDEBUG   -std=gnu++11 -o CMakeFiles/project.dir/component.cpp.o -c ~/git/project/subdir/component/component.cpp"` (home directory replaced manually for github)

In ~/foo/include I have (among others) 2 files:
- foo/class1.hpp
- foo/class2.hpp (includes class1)

All headers use `#pragma once`, class1&2 are in a namespace `foo::bar` (no alias).

In `component.cpp` I have multiple includes like `#include <foo/class2.hpp>`. I can `goto` class1, although I did not include its header. But I cannot `goto` class2.

I tried adding the include to class1 just above the class2 include and noticed something strange: CTRL+click on class1 include opens the class2 header. So it seams that some internal database is outdated. Although the goto behavior is unchanged.

Anything else I could do to gather more input?

**Comment by bobbrow** [comment]
@bobbrow said:

Things I can rule out:
* compile_commands.json does not currently affect goto definition.
* `#include` does not affect goto definition (unless the header lives outside of `${workspaceFolder}`). Symbols are found based on the value of `browse.path` (or `includePath` if `browse.path` is not present).

In the case where you added the new `#include` line, did you save the file before CTL+click?  The DB is currently not updated until you save your changes so that might explain what happened.

If I had to make a guess, the problem is that class2 isn't in the database because the tag parser failed somewhere while processing the header.  You are compiling with `-std=gnu++11`, so maybe there's a GCC extension we don't handle...  Can you check by opening class2.hpp and putting the text cursor somewhere inside the definition of class2?  The status bar should show which class/function you're in.  If it says something wrong, like "(Global Scope)" or some other symbol, then that would confirm my suspicion.

**Comment by bobbrow** [comment]
@FlameFire said:

The headers of class1/2 are outside the workspaceFolder if this matters.

Yes I did save. I also get the out "Checking for syntax errors" message for that file, but only, when I switch the editor tab to another file (and not yet back)

I'm pretty sure there is no GCC extension used. `gnu++11` is just the default from CMake (we should change that). But you are right with the `(Global Scope)`. For class1 it works, for class2 it does not.

I tried replacing the whole file contents of class2.hpp by `class foo{   };` and saved, switched tabs etc. but could not get it to work.

**Comment by bobbrow** [comment]
@bobbrow said:

Headers outside the workspace are not parsed by default.  They will be parsed if you open them in the editor or another file includes them.

Ok, so now that we know there's a tag parser failure involved, you may be able to help us figure out the symbol that is causing the failure.  It's usually related to macros, but it doesn't have to be.  If you start from the top of class2's header file and move the cursor down line by line, you should be able to see the status bar's symbol location update when you are inside a class or function.  The location where it stops updating as expected is the approximate location of the problem.  Unfortunately, we don't turn the DB icon back on when we update the DB for modified files (because it's usually a pretty fast operation), so you might just wait a few seconds to ensure the file has been tag parsed before doing this.

**Comment by bobbrow** [comment]
@FlameFire said:

I have the feeling it is not parsed at all although it is included the the component.cpp. As I said: I replaced the whole contents of that header and it still failed to parse that. I'll try again on monday and wait longer than I did before but I doubt it will change anything. Other ideas? Is there debug output of the tag parser?
BTW: What do you mean with the DB icon?

**Comment by bobbrow** [comment]
@bobbrow said:

You can set `"C_Cpp.loggingLevel": "Debug"`, run the "Reset IntelliSense Database" command and then look in the Output window with the C/C++ filter applied.  You should see a bunch of "tag parsing file:" messages.  Search the output for your header.  As long as it is included by at least one file in the workspace, it should be parsed.

**Comment by bobbrow** [comment]
@FlameFire said:

Note: I had to restart VSC to have the log level have any effect. It seems I also had to explicitly use
```
    "C_Cpp.intelliSenseEngine": "Default",
    "C_Cpp.intelliSenseEngineFallback": "Enabled",
```
However this did not help:
- I had reopened the folder and my symbols (including navigation) worked
- Navigation to include files broken: Not working at all or e.g. `#include <foobar/event.hpp>` jumped into a boost header that had a `event* classmember_` line. Looks like the parser thinks it is a C++ symbol rather than a header
- Apply Debug and above settings and multiple restarts: navigation more or less completely broken. The header files in question don't show up in the log at all (although there are lots of `tag parsing` entries)

Some redacted log output: [vscode.log](https://github.com/Microsoft/vscode-cpptools/files/2197624/vscode.log)
Note: I see quite a few `Tag parsing encountered a error, but it may not matter. Let us know if symbols in the file can't be found: ` in that. Mostly boost, but some also user includes. Not the ones I'm looking for though.
Note2: Reset Intellisense does not retrigger a parse of my open cpp file. Doing a edit-save gives me:

> textDocument/didSave
> Checking for syntax errors: file:///<myfile.cpp>
> queue_update_intellisense for files in tu of: <myfile.cpp>
> sending 2 changes to server
> errorSquiggles count: 0
> Checking for syntax errors: file:///<myfile.cpp>
> queue_update_intellisense for files in tu of: <myfile.cpp>
> errorSquiggles count: 0

Edit: I switched to `C_Cpp.intelliSenseEngine": "Tag Parser",` and GoTo started working.

**Comment by bobbrow** [comment]
@bobbrow said:

> I had reopened the folder and my symbols (including navigation) worked

So the problem was fixed by reloading the window?

> Edit: I switched to C_Cpp.intelliSenseEngine": "Tag Parser", and GoTo started working.

That's strange... GotoDef shouldn't be affected by that setting.

I'm going to move this to a new issue since the current issue is about something else.

**Comment by Flamefire** [comment]
> So the problem was fixed by reloading the window?

Not sure if it was "fixed" or what exactly changed. I just had to reopen the window as I did close it on Friday. My guess is, that the tag parser took over and provided the symbols but I have nothing to really base that on.

>That's strange... GotoDef shouldn't be affected by that setting.

Well it seems that the `Default` IntelliSense is broken in a way that affects the tag parser. Are you sure, that GoToDef is based ONLY on the tag parser? Anything else I can provide to help you find this?

**Comment by Flamefire** [comment]
Sorry again, it seems to have been a mistake on my side although I could not find much documentation related to this: `${workspaceFolder}` must be in `browse.path` for this to work. I removed it and run into the same issue again.

However the inconsistency in the database could be considered a bug: Half of the stuff was working, the other stuff was not (probably everything after removing the `browse.path` setting). Also open files do not seem to get parsed when they are not in the `browse.path` which is also annoying at the least.

**Comment by bobbrow** [comment]
For files not included by `browse.path` we remove symbols from those files from your DB when those files are closed so as not to pollute your workspace's DB.  Is that what's happening?

**Comment by Flamefire** [comment]
Not exactly. To summarize what I experienced:
- component.cpp is in `workspaceFolder` but not in `browse.path`, it includes `class1.hpp` and `class2.hpp`
- class1/2.hpp are in `browse.path` (and even in the same folder)
- GoToDef was working in component.cpp to (probably) all local symbols and to class2.hpp but not to class1.hpp
- Adding additional includes to component.cpp made GoToDef for headers resolved the wrong headers (shifted by line)
- Rebuilding the DB did not resolve anything from neither include
- component.cpp was never closed
- opening/closing class1/2.hpp did not change anything

So from that it seems that files not in `browse.path` are not parsed at all and that there were even stale symbols from the time it was parsed last (probably before it was removed from `browse.path`). I'll have to double-check, if the symbols in component.cpp are working, when it is not in `browse.path` or if only the headers are not followed/parsed.

**Comment by bobbrow** [comment]
> Adding additional includes to component.cpp made GoToDef for headers resolved the wrong headers (shifted by line)

Did you save the file before doing GotoDef in this case?  The browse DB is not continuously updated as you type.  You must save component.cpp first.

> Rebuilding the DB did not resolve anything from neither include

I'm not following... I thought there was only a problem with class1.hpp. (though earlier in the thread it was class2 that had problems - are you having issues with both files or did the names just get swapped for this example?)

> component.cpp was never closed

> So from that it seems that files not in `browse.path` are not parsed at all and that there were even stale symbols from the time it was parsed last (probably before it was removed from `browse.path`). I'll have to double-check, if the symbols in component.cpp are working, when it is not in `browse.path` or if only the headers are not followed/parsed.

I just looked at the code and it turns out I was wrong about this.  If the file is not in the browse.path the symbols won't immediately be deleted on file close.  The symbols will stick around until the browse.path changes, the database is reset, or you close/reopen the folder.  But it seems that it's not the symbols in this file that are causing you trouble anyway.  The real problem (I think) is still that the tag parser is failing to parse class1.hpp.  But were you saying that setting `"C_Cpp.intelliSenseEngine": "Tag Parser"` fixes that problem?

**Comment by Flamefire** [comment]
> Did you save the file before doing GotoDef in this case? The browse DB is not continuously updated as you type. You must save component.cpp first.

Yes, You mentioned that earlier, so I made sure of that

> I'm not following... I thought there was only a problem with class1.hpp. (though earlier in the thread it was class2 that had problems - are you having issues with both files or did the names just get swapped for this example?)

Yes, sorry switched the names as they are made up anyway. Essence: Only one was working

> The real problem (I think) is still that the tag parser is failing to parse class1.hpp. But were you saying that setting "C_Cpp.intelliSenseEngine": "Tag Parser" fixes that problem?

Yes, that fixes it. And I do have a repro now once I identified the problem: [vscodeTest.zip](https://github.com/Microsoft/vscode-cpptools/files/2200657/vscodeTest.zip)

The unexpected part: This is related to the `compileCommands`!

- Extract, `cd build && cmake ../src -DCMAKE_EXPORT_COMPILE_COMMANDS=ON`
- Open src folder in VSC (already contains a `.vscode`)
- Adjust the hard coded paths in c_cpp_properties.json
- play around with `browse.path` (Make sure to always Reset the DB and change+save the cpp): 

Results:

- `${workspaceFolder}` + `/.../include`: Everything works (GotToDef to class1/2 and in the file)
- Either one or both are missing: Nothing works (not even in the file)
- `/.../include` missing: Message from debug: `class1/2.hpp removed` (Just FYI)
- Remove `${workspaceFolder}` -> Nothing working, additionally remove `compileCommands` -> Everything works

**Comment by bobbrow** [comment]
Thanks for the repro project; that made my life much easier.  I'm going to debug this and try to get a fix out for the next release.

**Comment by bobbrow** [comment]
Fix will be in 0.17.7

**Comment by sean-mcmanus** [comment]
@Flamefire The Go to Definition problems have been fixed with 0.17.7.
